File: packages/protocol/src/consensus.ts
import type { ConsensusLabel } from "./schemas/common.js";
import type { Ballot } from "./schemas/vote.js";

// M0 fix (risk #3): 原设计用 "3/3 approve" 这类硬编码计数判定共识，
// 模型数一变就要重写。这里改成比例阈值，支持任意数量的模型。

export interface ConsensusThresholds {
  /** approve_ratio 达到此值才算 strong_consensus（默认要求全体 approve）。 */
  strongApproveRatio: number;
  /** approve_ratio 达到此值且无 critical 反对，算 qualified_consensus。 */
  qualifiedApproveRatio: number;
  /** approve_ratio 低于或等于此值，直接 rejected。 */
  rejectApproveRatio: number;
}

export const DEFAULT_CONSENSUS_THRESHOLDS: ConsensusThresholds = {
  strongApproveRatio: 1.0,
  qualifiedApproveRatio: 0.66,
  rejectApproveRatio: 0.34,
};

export interface ClassifyCandidateInput {
  ballotsForCandidate: Ballot[];
  /** 配置的模型数量（法定投票人数），不是实际投票数——用于计算比例和识别 partial。 */
  expectedVoterCount: number;
  thresholds?: ConsensusThresholds;
}

export interface ClassifyCandidateResult {
  label: ConsensusLabel;
  approveRatio: number;
  hasCriticalObjection: boolean;
  hasMajorObjection: boolean;
  /** 实际投票数少于 expectedVoterCount，说明有模型未响应（quorum 相关）。 */
  partial: boolean;
}

export function classifyCandidate(
  input: ClassifyCandidateInput
): ClassifyCandidateResult {
  const { ballotsForCandidate, expectedVoterCount } = input;
  const thresholds = input.thresholds ?? DEFAULT_CONSENSUS_THRESHOLDS;

  if (expectedVoterCount <= 0) {
    throw new Error("expectedVoterCount must be > 0");
  }

  const approveCount = ballotsForCandidate.filter(
    (b) => b.vote === "approve" || b.vote === "approve_with_conditions"
  ).length;
  const approveRatio = approveCount / expectedVoterCount;

  const objections = ballotsForCandidate.filter((b) => b.vote === "object");
  const hasCriticalObjection = objections.some(
    (b) => b.objection_severity === "critical"
  );
  const hasMajorObjection = objections.some(
    (b) => b.objection_severity === "major"
  );
  const partial = ballotsForCandidate.length < expectedVoterCount;

  let label: ConsensusLabel;
  if (hasCriticalObjection) {
    // critical objection 直接进入 disputed，不能被多数票吞掉。
    label = "disputed";
  } else if (hasMajorObjection) {
    label =
      approveRatio >= thresholds.qualifiedApproveRatio ? "disputed" : "rejected";
  } else if (approveRatio >= thresholds.strongApproveRatio) {
    label = "strong_consensus";
  } else if (approveRatio >= thresholds.qualifiedApproveRatio) {
    label = "qualified_consensus";
  } else if (approveRatio <= thresholds.rejectApproveRatio) {
    label = "rejected";
  } else {
    label = "disputed";
  }

  return { label, approveRatio, hasCriticalObjection, hasMajorObjection, partial };
}

File: python/mmd_litellm/consensus.py
from __future__ import annotations

from pydantic import BaseModel

from .schemas import Ballot, ConsensusLabel


class ConsensusThresholds(BaseModel):
    strong_approve_ratio: float = 1.0
    qualified_approve_ratio: float = 0.66
    reject_approve_ratio: float = 0.34


DEFAULT_CONSENSUS_THRESHOLDS = ConsensusThresholds()


class ClassifyCandidateResult(BaseModel):
    label: ConsensusLabel
    approve_ratio: float
    has_critical_objection: bool
    has_major_objection: bool
    partial: bool


def classify_candidate(
    ballots_for_candidate: list[Ballot],
    expected_voter_count: int,
    thresholds: ConsensusThresholds = DEFAULT_CONSENSUS_THRESHOLDS,
) -> ClassifyCandidateResult:
    if expected_voter_count <= 0:
        raise ValueError("expected_voter_count must be > 0")

    approve_count = sum(
        1
        for ballot in ballots_for_candidate
        if ballot.vote in ("approve", "approve_with_conditions")
    )
    approve_ratio = approve_count / expected_voter_count

    objections = [ballot for ballot in ballots_for_candidate if ballot.vote == "object"]
    has_critical_objection = any(
        ballot.objection_severity == "critical" for ballot in objections
    )
    has_major_objection = any(
        ballot.objection_severity == "major" for ballot in objections
    )
    partial = len(ballots_for_candidate) < expected_voter_count

    if has_critical_objection:
        label: ConsensusLabel = "disputed"
    elif has_major_objection:
        label = (
            "disputed"
            if approve_ratio >= thresholds.qualified_approve_ratio
            else "rejected"
        )
    elif approve_ratio >= thresholds.strong_approve_ratio:
        label = "strong_consensus"
    elif approve_ratio >= thresholds.qualified_approve_ratio:
        label = "qualified_consensus"
    elif approve_ratio <= thresholds.reject_approve_ratio:
        label = "rejected"
    else:
        label = "disputed"

    return ClassifyCandidateResult(
        label=label,
        approve_ratio=approve_ratio,
        has_critical_objection=has_critical_objection,
        has_major_objection=has_major_objection,
        partial=partial,
    )

